app/src/main/java/com/github/h3lp3rs/h3lp/forum/FireForum.kt

```diff
@@ -0,0 +1,262 @@
+            // we thus add it to the posts in the database (as explained in the forum protocol)
+            if (isCategory()) {
+                rootForum.addToObjectsListConcurrently(
+                    pathToKey(path + POSTS_LIST),
+                    String::class.java,
+                    key
+                )
+            }
+            // We can't use getPost here since we aren't sure that the setObject succeeded yet
+            ForumPost(this, forumPostData, emptyList())
+        }
+    }
+
+    override fun getPost(relativePath: Path): CompletableFuture<ForumPost> {
+        val fullPath = path + relativePath
+        val key = pathToKey(fullPath)
+
+        // Get the required post
+        val postFuture = rootForum.getObject(key, ForumPostData::class.java)
+
+        // Also get that post's replies
+        val postAndRepliesFuture = postFuture.thenCompose { postData ->
+            // Get all the replies from that post
+            rootForum.getObjectsList(
+                pathToKey(path + postData.repliesKey),
+                ForumPostData::class.java
+            ).handle { replies, error ->
+                if (error != null) {
+                    // If the post has no replies yet
+                    ForumPost(parent(), postData, emptyList())
+                } else {
+                    // If the post had replies, add them to the object
+                    ForumPost(parent(), postData, replies)
+                }
+            }
+        }
+        return postAndRepliesFuture
+    }
+
+    override fun getAll(): CompletableFuture<List<CategoryPosts>> {
+        if (isRoot()) {
+            // In case we are in the root forum, get the CategoryPosts from all categories
+            var future: CompletableFuture<List<CategoryPosts>> =
+                CompletableFuture.completedFuture(emptyList())
+            for (category in values()) {
+                // For all categories, we add them to the list of category posts
+                val categoryForum = forumOf(category)
+                future = future.thenCompose { list ->
+                    // Recursively call getAll
+                    categoryForum.getAll().handle { it, error ->
+                        if (error != null) {
+                            // If there is no posts on this forum, keep the current list of
+                            // category posts
+                            list
+                        } else {
+                            list + it
+                        }
+                    }
+                }
+            }
+            return future
+        } else {
+            // In case we are in a category, we only have a single CategoryPost to return (thus
+            // we return a singleton list)
+            return getAllFromCategory().thenApply { listOf(it) }
+        }
+    }
+
+    private fun getAllFromCategory(): CompletableFuture<CategoryPosts> {
```

Comment:
Eventhough it is private it could have some function description for understanding purposes

Reply:
> Really Nice code! The interface is really good documented and seems really pleasant to use. great job.
> 
> Maybe just one or two thing to say:
> 
>     1. Is it a good idea to have redundancy between MedicalType and ForumCategory. Wouldn't that cause some difficulty when linking them together. Maybe we can now remplace medicalType with a list of ForumCategory.
> 
>     2. It seems that the profil.png image has already changed color from teal to persimmon. (Probably my fault but dont know how it happened). It will be better for the presentation if we have the teal version this week. Just roll back this file and it will be fine.

Thanks for the suggestions, they've all been implemented